obs-outputs: Add eRTMP/eFLV multitrack support

palana opened the pull request:
### Description
Add eRTMP multitrack video support for the rtmp output and "full" eFLV support for the FLV output (both additional video codecs and multitrack video)

### Motivation and Context
This allows transmitting multiple video streams within the same RTMP connection; it's used for Twitch Enhanced Broadcasting to allow multiple renditions to be sent through a single RTMP connection, helping ensure that renditions remain aligned

### How Has This Been Tested?
It's being used in the Twitch Enhanced Broadcasting Beta

### Types of changes
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented:
With https://github.com/obsproject/obs-studio/pull/10494 merged, this can be rebased and undrafted.

palana commented:
> With #10494 merged, this can be rebased and undrafted.

@RytoEX: done

RytoEX commented on plugins/obs-outputs/flv-mux.c:
```diff
@@ -371,30 +389,46 @@ void flv_packet_ex(struct encoder_packet *packet, enum video_id_t codec_id,
+		s_w4cc(&s, codec_id);
+	}
+
+	// h264/hevc composition time offset
```
```suggestion
	// H.264/HEVC composition time offset
```

RytoEX commented on plugins/obs-outputs/flv-output.c:
```diff
@@ -41,12 +46,101 @@ struct flv_output {
+/* clang-format off */
+
+/**
+  * Chromaticity coordinates of the source primaries.
```
```suggestion
  * Chromaticity coordinates of the source primaries
```
or
```suggestion
  * Define chromaticity coordinates of the source primaries
```

The first is more "label-like" or "header-like" and would be appropriate here.

RytoEX commented on plugins/obs-outputs/flv-output.c:
```diff
@@ -41,12 +46,101 @@ struct flv_output {
+};
+
+/**
+ * Color Transfer Characteristic.
```
```suggestion
 * Color Transfer Characteristic
```

RytoEX commented on plugins/obs-outputs/flv-output.c:
```diff
@@ -41,12 +46,101 @@ struct flv_output {
+};
+
+/**
+ * YUV colorspace type.
```
```suggestion
 * YUV colorspace type
```

RytoEX commented on plugins/obs-outputs/flv-output.c:
```diff
@@ -111,42 +236,211 @@ static void write_meta_data(struct flv_output *stream)
+
+	case CODEC_H264:
+		packet.size = obs_parse_avc_header(&packet.data, header, size);
+		// Always send H264 on track 0 as old style for compat
```
```suggestion
		// Always send H.264 on track 0 as old style for compatibility.
```

RytoEX reviewed (changes requested):
Nits.

PatTheMav commented on plugins/obs-outputs/flv-mux.c:
```diff
@@ -59,22 +66,32 @@ enum datatype_t {
 {
 	switch (id) {
+	case CODEC_NONE:
+		assert(0 && "Tried to serialize CODEC_NONE");
```
We set `NDEBUG` for all configurations apart from `Debug`, so the assert is inert in those. Is that taken into account here?

RytoEX replied:
I briefly checked with @notr1ch off-thread and it seems that assert here is okay. @palana could you summarize the intent with the assert here to make sure we're all on the same page?

RytoEX replied:
This is fine as-is for now then. If need be, we'll re-evaluate later.

Lain-B reviewed (approved):
Outside of the ryan/pat nitpicks this has my approval. An alternative for the assert might be like, bcrash() or something. (If it's not just a debug situation at least)

palana commented:
@RytoEX: nits should be addressed and squashed (and rebased)

derrod commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -922,33 +961,41 @@ static inline bool send_headers(struct rtmp_stream *stream)
 
-	// Y2023 spec
-	return send_video_footer(stream);
+		if (stream->video_codec[i] == CODEC_H264)
+			continue;
```
If H.264 is not on track 0, i.e. sent as "enhanced" packet, it should also send a footer.

palana replied:
changed in https://github.com/obsproject/obs-studio/pull/10500#commits-pushed-bac01a2

derrod commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -922,33 +961,41 @@ static inline bool send_headers(struct rtmp_stream *stream)
 			return false;
 	}
 
+	i = 1;
+	while (obs_output_get_video_encoder2(stream->output, i) != NULL) {
+		if (!send_video_metadata(stream, i) ||
+		    !send_video_header(stream, i))
+			return false;
+		i += 1;
+	}
+
```
Is there any situation where there could be a gap in encoders? i.e. 0, 1, 3 are valid but 2 is empty?

Either way I'd personally write this as a for-loop:
```c
for (size_t idx = 1; idx < MAX_OUTPUT_VIDEO_ENCODERS; idx++) {
    obs_encoder_t *enc = obs_output_get_video_encoder2(stream->output, idx);
    if (!enc)
        continue;
    if (!send_video_metadata(stream, idx) || !send_video_header(stream, idx))
        return false;
}
```

palana replied:
the beta multitrack video output doesn't create gaps, but there shouldn't be a reason to not support them

palana replied:
@derrod: might be interesting to also change this for audio tracks

palana replied:
changed in https://github.com/obsproject/obs-studio/pull/10500#event-12641499492

derrod commented on plugins/obs-outputs/rtmp-stream.c:
```diff
@@ -922,33 +961,41 @@ static inline bool send_headers(struct rtmp_stream *stream)
-			return false;
-
-	if (!send_video_header(stream))
+	if (!send_video_metadata(stream, 0) || !send_video_header(stream, 0))
```
Also in general, is there a reason this is split from the metadata for idx > 0? Why not send all of them in one big loop?

palana replied:
not as far as I'm aware, no; @kc5nra: do you know if there's a reason to split headers/metadata like this?

RytoEX replied:
This feedback thread is still active.

palana replied:
changed in https://github.com/obsproject/obs-studio/pull/10500#commits-pushed-cf791ae

derrod reviewed (changes requested):
A couple smaller things, it would be good if you could also adjust `init_send()` to correctly accumulate bitrate from all video encoders rather than just the primary one, that should fix the issues peopole have seen with the network enhancement options, but that's not a blocker for now.

RytoEX reviewed (changes requested):
Nits. I'll probably pass after this round.

RytoEX commented on plugins/obs-outputs/flv-mux.c:
```diff
@@ -371,30 +389,46 @@ void flv_packet_ex(struct encoder_packet *packet, enum video_id_t codec_id,
-	s_w8(&s,
-	     FRAME_HEADER_EX | type | (packet->keyframe ? FT_KEY : FT_INTER));
-	s_w4cc(&s, codec_id);
+	s_w8(&s, FRAME_HEADER_EX |
+			 (is_multitrack ? PACKETTYPE_MULTITRACK : type) |
+			 (packet->keyframe ? FT_KEY : FT_INTER));
```
This feedback thread is still active.

palana replied:
changed in https://github.com/obsproject/obs-studio/pull/10500#commits-pushed-cf791ae

RytoEX commented:
Off-thread discussion indicates that this looks good. Squash the commits appropriately and we'll give it a (hopefully) final pass.

palana commented:
> Off-thread discussion indicates that this looks good. Squash the commits appropriately and we'll give it a (hopefully) final pass.

@RytoEX: squashed and rebased
